File: 02/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type game struct {
	strategy   rune
	theirs rune
}

func gameFromStr(input string) game {
	splitInput := []rune(input)
	return game{strategy: splitInput[2], theirs: splitInput[0]}
}

func gamesFromInput() []game {
	inputFile, _ := os.Open("input.txt")
	defer inputFile.Close()

	lines := bufio.NewScanner(inputFile)

	games := []game{}
	for lines.Scan() {
		g := gameFromStr(lines.Text())

		games = append(games, g)
	}

	return games
}

func scoreFrom(g game) int {
	outcomes := map[game]int{
		game{'X', 'A'}: 0 + 3,
		game{'X', 'B'}: 0 + 1,
		game{'X', 'C'}: 0 + 2,
		game{'Y', 'A'}: 3 + 1,
		game{'Y', 'B'}: 3 + 2,
		game{'Y', 'C'}: 3 + 3,
		game{'Z', 'A'}: 6 + 2,
		game{'Z', 'B'}: 6 + 3,
		game{'Z', 'C'}: 6 + 1,
	}

	return outcomes[g]
}

func main() {
	games := gamesFromInput()

	score := 0
	for _, currentGame := range games {
		score += scoreFrom(currentGame)
	}

	fmt.Println(score)
}

File: 06/part2.py
def firstMarkerPosFor(line):
    for idx, char in enumerate(line):
        if startOfMessage(idx, line):
            return idx + 1

    return None


def startOfMessage(idx, line):
    if idx < 14:
        return False

    checkedChars = set()

    for i in range(14):
        if line[idx-i] in checkedChars:
            return False

        checkedChars.add(line[idx-i])

    return True


with open('input.txt') as file:
    for line in file:
        print(firstMarkerPosFor(line))
